openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -223,7 +223,9 @@ def is_flag_active(self, flag_name, check_before_waffle_callback=None, flag_unde
                 if value is None:
                     value = flag_is_active(get_request(), namespaced_flag_name)
 
-            self._cached_flags[namespaced_flag_name] = value
+            # Don't cache the result if it might have been affected by the callback
```

Comment:
Thanks @robrap. I realized while walking my dog that this approach doesn't work. Here are some further thoughts:
1. Not caching if the callback is specified means the code essentially never caches any more, since we always use CourseWaffleFlag.
2. The reason we haven't seen the problem before is the problem occurs if the same flag is queried with multiple keys in one request. For most pages, all requests will be for the same course. The dashboard is the first place where we've had two courses queried within a single request.
3. The important thing is that we minimize the number of database queries. The behavior we need is that we only look up each record once, so I think we need separate caches for the Waffle flag and for the override table.

I'm thinking of removing this commit from this PR so that we can merge it. I'll make a JIRA ticket for the bug which we should fix before we start to use this flag in production. We need to write a test where we query across multiple courses within a test. We should probably also write a query verification test that we can make many calls across multiple courses and only have a minimal number of queries made.